File: ConsoleApp1/ConsoleApp1/DocGia.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace QLTV
{
    class DocGia : Nguoi, DacTrungDocGia
    {
        private int iMaDG;
        private string sEmail;
        private Date dtNgayHetHanThe;
        private double dTaiKhoan;
		public int MaDG
		{
			get
			{
				return this.iMaDG;
			}
			set
			{
				this.iMaDG = value;
			}
		}
		public string Email
		{
			get
			{
				return this.sEmail;
			}
		}
		public Date NgayHetHanThe
		{
			get { return this.dtNgayHetHanThe; }
			set { this.dtNgayHetHanThe = value; }
		}
		public double TaiKhoan
		{
			get { return this.dTaiKhoan; }
			set
			{
				if (value < 0 )
					throw new ArgumentOutOfRangeException
					(" must be > 0");
				this.dTaiKhoan = value;
			}
		}
		public DocGia() : base()
        {
			this.dtNgayHetHanThe = new Date();
		}
        public DocGia(int MaDG, string HoTen, string DiaChi, string SDT, int NgaySinh,int ThangSinh, int NamSinh, string Email, int NgayHH, int ThangHH, int NamHH, double TaiKhoan) : base(HoTen, DiaChi, SDT, NgaySinh,ThangSinh,NamSinh)
        {
            this.iMaDG = MaDG;
            this.sEmail = Email;
            this.dtNgayHetHanThe = new Date();
            this.dtNgayHetHanThe.Ngay = NgayHH;
            this.dtNgayHetHanThe.Thang = ThangHH;
            this.dtNgayHetHanThe.Nam = NamHH;
            this.dTaiKhoan = TaiKhoan;
        }
		public override void Nhap()
		{
			base.Nhap();
			Console.Write("Email: ");
			this.sEmail = Console.ReadLine();
		}
        public override void Xuat()
        {
            base.Xuat();
            Console.WriteLine("Ma doc gia: " + this.iMaDG);
            Console.WriteLine("Email: " + this.sEmail);
            Console.WriteLine("Ngay het han the: " + this.dtNgayHetHanThe.Ngay+"/"+this.dtNgayHetHanThe.Thang+"/"+this.dtNgayHetHanThe.Nam);
            Console.WriteLine("So du tai khoan: " + this.dTaiKhoan);
        }
		public void GiaHanThe(Date Today)
		{
			if (Today < this.dtNgayHetHanThe)
			{
				this.dtNgayHetHanThe += 180;
				this.dTaiKhoan = this.dTaiKhoan - 50000;
			}
			else
			{
				this.dtNgayHetHanThe = Today + 180;
				this.dTaiKhoan = this.dTaiKhoan - 50000;
			}
		}
		public override void Tien(double Tien)
		{
			this.dTaiKhoan = this.dTaiKhoan + Tien;
		}
		public void DangKyTaiKhoanMoi(Date Today)
		{
			this.dtNgayHetHanThe = Today + 180;
			this.dTaiKhoan = 100000;
		}
	}
}

File: Quan_Ly_Thu_Vien/src/DocGia.java
import java.io.IOException;
import java.util.Scanner;

public class DocGia extends Nguoi implements DacTrungDocGia 
{
	private int iMaDG;
	private String sEmail;
	private Date dtNgayHetHan;
	private double dTaiKhoan;
	
	public int getiMaDG() 
	{
		return iMaDG;
	}
	public void setiMaDG(int iMaDG)
	{
		this.iMaDG = iMaDG;
	}
	public String getsEmail()
	{
		return sEmail;
	}

	public void setsEmail(String sEmail)
	{
		this.sEmail = sEmail;
	}
	public double getdTaiKhoan() 
	{
		return this.dTaiKhoan;
	}

	public void setdTaiKhoan(double dTaiKhoan) 
	{
		this.dTaiKhoan = dTaiKhoan;
	}
	public Date getdtNgayHetHan() 
	{
		return dtNgayHetHan;
	}
	public void setdtNgayHetHan(Date dtNgayHetHan)
	{
		this.dtNgayHetHan = dtNgayHetHan;
	}
	
	public DocGia() 
	{
		super();
		this.dtNgayHetHan= new Date();
	}
	
	
	public DocGia (int MaDG, String HoTen, String DiaChi, int SDT, String Email, int NgayHH,int ThangHH,int NamHH, int TKSV)
	{
		super(HoTen, DiaChi, SDT);
		this.iMaDG=MaDG;
		this.sEmail= Email;
		this.dtNgayHetHan= new Date();
		this.dtNgayHetHan.setdNgay(NgayHH);
		this.dtNgayHetHan.setdThang(ThangHH);
		this.dtNgayHetHan.setdNam(NamHH);
		this.dTaiKhoan = TKSV;
	}

	@SuppressWarnings("resource")
	@Override
	public void Nhap() throws IOException 
	{
		super.Nhap();
		Scanner sc = new Scanner(System.in);
		System.out.println("Nhap  vao Email: ");
		this.sEmail = sc.nextLine();
	}
	@Override
	public void Xuat() throws IOException
	{
		super.Xuat();
		System.out.println(" Ma doc gia: "+this.iMaDG);
		System.out.println(" Email doc gia: "+this.sEmail);
		System.out.println(" Ngay het han the la: ");
		System.out.println(" Ngay: " + this.dtNgayHetHan.getiNgay());
		System.out.println(" Thang: " + this.dtNgayHetHan.getiThang());
		System.out.println(" Nam: " + this.dtNgayHetHan.getiNam());
		System.out.println(" So du tai khoan; "+this.dTaiKhoan);
		
	}
	@Override
	public void GiaHanTheThuVien(Date Today) 
	{
		if (Date.SoSanhNhoHon(Today,this.dtNgayHetHan)==true)
		{
			this.dtNgayHetHan = Date.PlusDay(this.dtNgayHetHan, 180);
			this.dTaiKhoan = this.dTaiKhoan - 50000;
		}
		else
		{
			this.dtNgayHetHan = Date.PlusDay(Today,180);
			this.dTaiKhoan = this.dTaiKhoan - 50000;
		}
	}
	@Override
	public void Tien(double Tien) 
	{
		this.dTaiKhoan = this.dTaiKhoan + Tien;
	}
	@Override
	public void DangKyTaiKhoanMoi(Date Today)
	{
		this.dtNgayHetHan = Date.PlusDay(Today,180);
		this.dTaiKhoan = 100000;
	}
}
